src/serializer/serializer.js

```diff
@@ -1192,29 +1196,53 @@ export class Serializer {
 
     let requireStatistics = { replaced: 0, count: 0 };
 
-    // Ensure that all bindings that actually get modified get proper variables
     let functionEntries: Array<[BabelNodeBlockStatement, FunctionInfo]> = Array.from(this.functions.entries());
-    for (let [, { instances, names }] of functionEntries) {
-      for (let instance of instances) {
-        let serializedBindings = instance.serializedBindings;
-        for (let name in names) {
-          let serializedBinding: SerializedBinding = serializedBindings[name];
-          if (serializedBinding.modified && !serializedBinding.referentialized) {
+
+    // Ensure that all bindings that actually get modified get proper variables
+    function allocateModifiedBindingVariables(instance: FunctionInstance, names, captureScope: boolean) {
+      let capturedScopeProperties : Array<BabelNodeObjectProperty> = [] ;
+
+      let serializedBindings = instance.serializedBindings;
+      for (let name in names) {
+        let serializedBinding : SerializedBinding = serializedBindings[name];
+        if (serializedBinding.modified && !serializedBinding.referentialized) {
+          if (captureScope) { // Initialize captured scope at function call instead of globally
+	    let property : BabelNodeObjectProperty = t.objectProperty(t.identifier(name), serializedBinding.serializedValue);
+            capturedScopeProperties.push(property);
+            serializedBinding.serializedValue = t.memberExpression(t.identifier(this.localCapturedScopeName),  t.identifier(name), false);
+          } else {
             let serializedBindingId = t.identifier(this.referentializedNameGenerator.generate(name));
             let declar = t.variableDeclaration("var", [
               t.variableDeclarator(serializedBindingId, serializedBinding.serializedValue)]);
             getFunctionBody(instance).push(declar);
             serializedBinding.serializedValue = serializedBindingId;
-            serializedBinding.referentialized = true;
-            this.statistics.referentialized++;
           }
+          serializedBinding.referentialized = true;
+          this.statistics.referentialized++;
         }
       }
+
+      if (captureScope && capturedScopeProperties.length) {
```

Comment:
Adding a comment at the top to illustrate what you're trying to build would be helpful.